cpp/cmake_modules/ThirdpartyToolchain.cmake

```diff
@@ -901,602 +992,876 @@ if (ARROW_JEMALLOC)
-    set(BROTLI_STATIC_LIBRARY_COMMON "${BROTLI_PREFIX}/${BROTLI_LIB_DIR}/${CMAKE_STATIC_LIBRARY_PREFIX}brotlicommon${CMAKE_STATIC_LIBRARY_SUFFIX}")
-    set(BROTLI_CMAKE_ARGS ${EP_COMMON_CMAKE_ARGS}
-      "-DCMAKE_INSTALL_PREFIX=${BROTLI_PREFIX}"
-      -DCMAKE_INSTALL_LIBDIR=lib
-      -DBUILD_SHARED_LIBS=OFF)
-
-    ExternalProject_Add(brotli_ep
-      URL ${BROTLI_SOURCE_URL}
-      BUILD_BYPRODUCTS "${BROTLI_STATIC_LIBRARY_ENC}" "${BROTLI_STATIC_LIBRARY_DEC}" "${BROTLI_STATIC_LIBRARY_COMMON}"
-      ${BROTLI_BUILD_BYPRODUCTS}
-      ${EP_LOG_OPTIONS}
-      CMAKE_ARGS ${BROTLI_CMAKE_ARGS}
-      STEP_TARGETS headers_copy)
-    if (MSVC)
-      ExternalProject_Get_Property(brotli_ep SOURCE_DIR)
+  # The include directory must exist before it is referenced by a target.
+  file(MAKE_DIRECTORY "${GBENCHMARK_INCLUDE_DIR}")
 
-      ExternalProject_Add_Step(brotli_ep headers_copy
-        COMMAND xcopy /E /I include ..\\..\\..\\brotli_ep\\src\\brotli_ep-install\\include /Y
-        DEPENDEES build
-        WORKING_DIRECTORY ${SOURCE_DIR})
-    endif()
-    set(BROTLI_VENDORED 1)
+  add_library(benchmark::benchmark STATIC IMPORTED)
+  set_target_properties(benchmark::benchmark PROPERTIES
+          IMPORTED_LOCATION "${GBENCHMARK_STATIC_LIB}"
+          INTERFACE_INCLUDE_DIRECTORIES "${GBENCHMARK_INCLUDE_DIR}"
+  )
 
-    add_dependencies(toolchain brotli_ep)
-  else()
-    find_package(Brotli REQUIRED)
-    set(BROTLI_VENDORED 0)
+  add_library(benchmark::benchmark_main STATIC IMPORTED)
+  set_target_properties(benchmark::benchmark_main PROPERTIES
+          IMPORTED_LOCATION "${GBENCHMARK_MAIN_STATIC_LIB}"
+          INTERFACE_INCLUDE_DIRECTORIES "${GBENCHMARK_INCLUDE_DIR}"
+  )
+
+  add_dependencies(toolchain-benchmarks gbenchmark_ep)
+  add_dependencies(benchmark::benchmark gbenchmark_ep)
+  add_dependencies(benchmark::benchmark_main gbenchmark_ep)
+endmacro()
+
+if(ARROW_BUILD_BENCHMARKS)
+  resolve_dependency(benchmark)
+  # TODO: Don't use global includes but rather target_include_directories
+  get_target_property(BENCHMARK_INCLUDE_DIR benchmark::benchmark INTERFACE_INCLUDE_DIRECTORIES)
+  include_directories(SYSTEM ${BENCHMARK_INCLUDE_DIR})
+endif()
+
+macro(build_rapidjson)
+  message(STATUS "Building rapidjson from source")
+  set(RAPIDJSON_PREFIX "${CMAKE_CURRENT_BINARY_DIR}/rapidjson_ep/src/rapidjson_ep-install")
+  set(RAPIDJSON_CMAKE_ARGS
+    -DRAPIDJSON_BUILD_DOC=OFF
+    -DRAPIDJSON_BUILD_EXAMPLES=OFF
+    -DRAPIDJSON_BUILD_TESTS=OFF
+    "-DCMAKE_INSTALL_PREFIX=${RAPIDJSON_PREFIX}")
+
+  ExternalProject_Add(rapidjson_ep
+    ${EP_LOG_OPTIONS}
+    PREFIX "${CMAKE_BINARY_DIR}"
+    URL ${RAPIDJSON_SOURCE_URL}
+    CMAKE_ARGS ${RAPIDJSON_CMAKE_ARGS})
+
+  set(RAPIDJSON_INCLUDE_DIR "${RAPIDJSON_PREFIX}/include")
+
+  add_dependencies(toolchain rapidjson_ep)
+  add_dependencies(rapidjson rapidjson_ep)
+endmacro()
+
+# TODO: Check for 1.1.0+
+if(ARROW_WITH_RAPIDJSON)
+  get_dependency_source(RapidJSON)
```

Comment:
Something odd is happening, it uses another build directory downloaded rapidjson:

```
-- RapidJSON found. Headers: /home/fsaintjacques/src/db/arrow/cpp/flight-build/rapidjson_ep/src/rapidjson_ep-install/include
```